### PR TITLE
add ACR repository operations

### PR DESCRIPTION
**Pull request by VeryEarly**
## Description

design: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/811

## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Accounts/Accounts/Context/ClearAzureRmContext.cs [hunk 1]
```diff
@@ -84,6 +87,15 @@ void ClearContext(AzureRmProfile profile, RMProfileClient client)
                     profile.TrySetDefaultContext(defaultContext);
                     result = true;
                 }
+
```
Comment (VeryEarly) [inline src/Accounts/Accounts/Context/ClearAzureRmContext.cs:90]: Maintain a list of registered shared components in common.
Clear shared components in one shot.
maybe?

File: src/ContainerRegistry/ContainerRegistry/Models/ContainerRegistryDataPlaneClient.cs [hunk 2]
```diff
@@ -14,46 +14,100 @@
...
+
+        public string TryAuthenticate(string scope = _scopeDefault)
+        {
+            _accessToken = TryGetToken(scope);
+            return _accessToken;
+        }
+
+        private string TryGetToken(string key)
+        {
+            IDictionary<string, Tuple<string, DateTime>> cache = TryGetTokenCache();
```
Comment (isra-fel) [inline src/ContainerRegistry/ContainerRegistry/Models/ContainerRegistryDataPlaneClient.cs:84]: Suggest `using AcrTokenCache = IDictionary<string, Tuple<string, DateTime>>` to simplify code

**Comment by VeryEarly** [comment]
CI should fail because no updated common module

File: src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Web;
+
+namespace Microsoft.Azure.Commands.ContainerRegistry.DataPlaneOperations
+{
+    public abstract class ContainerRegistryDataPlaneOperationBase<T>
+    {
+        protected ContainerRegistryDataPlaneClient _client;
```
Comment (VeryEarly) [inline src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs:22]: remove underscore, use pascal case

File: src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        protected virtual string _scope
```
Comment (VeryEarly) [inline src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs:23]: remove underscore, use pascal case

File: src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            this._client = client;
+        }
+
+        public T ProcessRequest()
+        {
+            _client.Authenticate(_scope);
+            return SendRequest();
+        }
+
+        public virtual T SendRequest()
```
Comment (VeryEarly) [inline src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs:70]: change virtual to abstract

File: src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            throw new NotImplementedException();
+        }
+
+        //sample link: </acr/v1/_catalog?last=test%2Fbusybox99&n=100&orderby=>; rel="next"
+        //repository should be in between of first appeartance of '=' and '&'
+        protected string getLastListed(string nextLink)
```
Comment (VeryEarly) [inline src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs:77]: capital G for GetLastListed

File: src/ContainerRegistry/ContainerRegistry/Models/ContainerRegistryDataPlaneClient.cs [hunk 7]
```diff
@@ -14,46 +14,84 @@
...
+                try
+                {
+                    value = new AcrToken(token);
+                }
+                catch
+                {
+                    throw new InvalidOperationException(string.Format("Invalud token for {0}", key));
+                }
+
+                cache.Add(key, value);
```
Comment (VeryEarly) [inline src/ContainerRegistry/ContainerRegistry/Models/ContainerRegistryDataPlaneClient.cs:92]: Change method name Add to Set

File: src/ContainerRegistry/ContainerRegistry/Models/AcrTokenCache.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            _cache.Clear();
+        }
+
+        public bool TryGetToken(string key, out AcrToken value)
+        {
+            return _cache.TryGetValue(key, out value);
+        }
+
+        public void Add(string key, AcrToken token)
```
Comment (VeryEarly) [inline src/ContainerRegistry/ContainerRegistry/Models/AcrTokenCache.cs:28]: change add to set

File: src/ContainerRegistry/ContainerRegistry/Models/AcrToken.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        public string GetToken()
+        {
+            return _token;
+        }
+
+        public bool IsExpired(int minutesBeforeExpiration)
+        {
+            return (_exp - DateTime.UtcNow).Minutes <= minutesBeforeExpiration;
```
Comment (VeryEarly) [inline src/ContainerRegistry/ContainerRegistry/Models/AcrToken.cs:33]: DateTime.UtcNow + minutesBeforeExpiration >= _exp

File: tools/StaticAnalysis/Exceptions/Az.ContainerRegistry/MissingAssemblies.csv [hunk 10]
```diff
@@ -1,4 +1,6 @@
...
-"Az.ContainerRegistry","System.IdentityModel.Tokens.Jwt","5.1.2.0","Microsoft.Azure.ContainerRegistry","0","3000","Missing assembly System.IdentityModel.Tokens.Jwt referenced from Microsoft.Azure.ContainerRegistry","Ensure that the assembly is included in the Wix file or directory"
\ No newline at end of file
+"Az.ContainerRegistry","System.IdentityModel.Tokens.Jwt","5.1.2.0","Microsoft.Azure.ContainerRegistry","0","3000","Missing assembly System.IdentityModel.Tokens.Jwt referenced from Microsoft.Azure.ContainerRegistry","Ensure that the assembly is included in the Wix file or directory"
+"Az.ContainerRegistry","System.Text.Encodings.Web","4.0.4.0","Azure.Identity","0","3000","Missing assembly System.Text.Encodings.Web referenced from Azure.Identity","Ensure that the assembly is included in the Wix file or directory"
+"Az.ContainerRegistry","System.Net.HttpListener","4.0.1.0","Microsoft.Identity.Client","0","3000","Missing assembly System.Net.HttpListener referenced from Microsoft.Identity.Client","Ensure that the assembly is included in the Wix file or directory"
```
Comment (isra-fel) [inline tools/StaticAnalysis/Exceptions/Az.ContainerRegistry/MissingAssemblies.csv]: Can we revert change to this file now?

File: src/ContainerRegistry/ContainerRegistry/ChangeLog.md [hunk 11]
```diff
@@ -18,6 +18,16 @@
         - Additional information about change #1
 -->
 ## Upcoming Release
+* Added cmdlets:
```
Comment (dingmeng-xue) [inline src/ContainerRegistry/ContainerRegistry/ChangeLog.md:21]: Supported repository, manifest, and tag operations
- cmdlet lists

Comment (dingmeng-xue) [reply src/ContainerRegistry/ContainerRegistry/ChangeLog.md:21]: Supported login to container registry through `Connect-AzContainerRegistry`

Comment (VeryEarly) [reply src/ContainerRegistry/ContainerRegistry/ChangeLog.md]: it was added from previous release

Comment (VeryEarly) [reply src/ContainerRegistry/ContainerRegistry/ChangeLog.md]: > Supported login to container registry through `Connect-AzContainerRegistry`

it was added in previous release

File: tools/StaticAnalysis/Exceptions/Az.ContainerRegistry/MissingAssemblies.csv (hunk 10 above)
Comment (VeryEarly) [reply tools/StaticAnalysis/Exceptions/Az.ContainerRegistry/MissingAssemblies.csv]: > Can we revert change to this file now?

done

File: src/ContainerRegistry/ContainerRegistry/help/Update-AzContainerRegistryTag.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+external help file: Microsoft.Azure.PowerShell.Cmdlets.ContainerRegistry.dll-Help.xml
+Module Name: Az.ContainerRegistry
+online version: https://docs.microsoft.com/en-us/powershell/module/az.containerregistry/update-azcontainerregistrytag
+schema: 2.0.0
+---
+
+# Update-AzContainerRegistryTag
+
+## SYNOPSIS
+Update Acr tag.
```
Comment (dingmeng-xue) [inline src/ContainerRegistry/ContainerRegistry/help/Update-AzContainerRegistryTag.md:11]: Recommended to use ACR

File: src/ContainerRegistry/ContainerRegistry/help/Az.ContainerRegistry.md [hunk 13]
```diff
@@ -71,6 +89,15 @@ Updates a container registry.
...
 Regenerates a login credential for a container registry.
 
+### [Update-AzContainerRegistryManifest](Update-AzContainerRegistryManifest.md)
+Update ACR manifest. 
+
+### [Update-AzContainerRegistryRepository](Update-AzContainerRegistryRepository.md)
+Update ACR repository.
+
+### [Update-AzContainerRegistryTag](Update-AzContainerRegistryTag.md)
+Update Acr tag.
```
Comment (dingmeng-xue) [inline src/ContainerRegistry/ContainerRegistry/help/Az.ContainerRegistry.md:99]: "ACR" should be consistent with others.

File: src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+        //sample link: </acr/v1/_catalog?last=test%2Fbusybox99&n=100&orderby=>; rel="next"
+        //repository should be in between of first appeartance of '=' and '&'
+        protected string GetLastListed(string nextLink)
+        {
+            int startOfParam = nextLink.IndexOf('<') + 1;
+            int endOfParam = nextLink.IndexOf('>') - startOfParam - 1;
+            string param = nextLink.Substring(startOfParam, endOfParam).Substring(nextLink.IndexOf('?') + 1);
+            string last = param.Split('&')[0].Split('=')[1];
+            return HttpUtility.UrlDecode(last);
```
[the remark this answers is not in the document]
Comment (VeryEarly) [reply src/ContainerRegistry/ContainerRegistry/DataPlaneOperations/ContainerRegistryDataPlaneOperationBase.cs:80]: This pattern is still not claimed anywhere and I would like to follow the current logic which was implement by service team themselves.

There are some swagger update needed, I'll ask them to specify this pattern and update this method accordingly then.

**Review by isra-fel — APPROVED** [review]

**Pull request merged**